### PR TITLE
feat: Allow building on Darwin for both arm64 and amd64 architectures

### LINKED ISSUE #30: Build for arm64 darwin
This issue tracks adding support for compiling KraftKit for darwin (macOS) arm64.

### PR DESCRIPTION
**Pull request by jake-ciolek**
### Prerequisite checklist

  - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
  - [x] Tested your changes against relevant architectures and platforms;
  - [x] Ran `make fmt` on your commit series before opening this PR;
  - [ ] Updated relevant documentation.

### Description of changes

Create darwin specific equivalents of _linux code.

If possible, merge common parts into _unix suffixed files. Move others into _linux.go and _darwin.go files.

Augment goreleaser's ytt templates to produce two additional targets:

- darwin amd64
- darwin arm64

Make it build a binary runnable on the host MacOS machine via make kraft (both with DOCKER variable set and not set)

GitHub-Fixes: #30

### CODE REVIEW

File: .goreleaser-stable.yaml [hunk 1]
```diff
@@ -57,20 +57,23 @@ nix:
       cp -vr ./dist/kraft $out/bin/kraft
 
 builds:
+#@ targets = {"linuxu": {"os": "linux", "arch": "amd64"}, "darwin-arm64": {"os": "darwin", "arch": "arm64"}, "darwin-amd64": {"os": "darwin", "arch": "amd64"}}
```
Comment (nderjung) [inline .goreleaser-stable.yaml]: I would turn this simply into tuples, it's also possible to use multiple lines using ytt:
```yaml
#@ targets = [ \
#@   ["linux", "amd64"], \
#@   ["linux", "arm64"], \
#@ ]
```
Should make git blames easier to read

File: .goreleaser-staging.yaml [hunk 2]
```diff
@@ -28,20 +28,24 @@ nfpms:
...
-      - amd64
+      - #@ specs["arch"]
     ldflags:
       - -s -w 
       - -X {{ .Env.GOMOD }}/internal/version.version={{ .Version }}
       - -X {{ .Env.GOMOD }}/internal/version.commit={{ .Commit }}
       - -X {{ .Env.GOMOD }}/internal/version.buildTime={{ .Date }}
 #@ end
+#@ end
+
```
Comment (craciunoiuc) [inline .goreleaser-staging.yaml]:
```suggestion
```

Extra space?

File: Makefile [hunk 3]
```diff
@@ -39,11 +39,14 @@ GIT_SHA     ?= $(shell git update-index -q --refresh && \
...
-               -e DOCKER= \
-               -w /go/src/$(GOMOD) \
-               -v $(WORKDIR):/go/src/$(GOMOD) \
-               $(REGISTRY)/$(2):$(IMAGE_TAG) \
-                 $(3)
+				-e DOCKER= \
+				-e GOOS=$(GOOS) \
+				-e GOARCH=$(GOARCH) \
+				-w /go/src/$(GOMOD) \
+				--platform linux/amd64 \
```
Comment (craciunoiuc) [inline Makefile]: Shouldn't the platform be based on the host?

File: machine/platform/register_darwin.go [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2022, Unikraft GmbH and The KraftKit Authors.
+// Licensed under the BSD-3-Clause License (the "License").
+// You may not use this file except in compliance with the License.
+package platform
+
+func unixVariantStrategies() map[Platform]*Strategy {
+	// Nothing added for Darwin
+	return map[Platform]*Strategy{}
+}
```
Comment (craciunoiuc) [inline machine/platform/register_darwin.go:13]: @nderjung If we get this PR in with stubs, I also want to get the FreeBSD PR in :D

File: machine/platform/register_linux.go [hunk 5]
```diff
@@ -69,13 +43,11 @@ var firecrackerV1alpha1Driver = func(ctx context.Context, opts ...any) (machinev
 	)
 }
 
-// hostSupportedStrategies returns the map of known supported drivers for the
-// given host.
-func hostSupportedStrategies() map[Platform]*Strategy {
+func unixVariantStrategies() map[Platform]*Strategy {
+	// (jake-ciolek): The firecracker driver has a dependency on github.com/containernetworking/plugins/pkg/ns via
```
Comment (craciunoiuc) [inline machine/platform/register_linux.go]:
```suggestion
	// TODO(jake-ciolek): The firecracker driver has a dependency on github.com/containernetworking/plugins/pkg/ns via
```

**Comment by nderjung** [comment]
Hi @jake-ciolek, thanks a lot for this work. I've turned #266 into an epic since it looks like we need several things to get this all tied up. Can you please remove the `GitHub-Fixes` from the commit message when you make your updates?

File: machine/platform/register_linux.go (hunk 5 above)
Comment (jake-ciolek) [reply machine/platform/register_linux.go]: Done

File: Makefile (hunk 3 above)
Comment (jake-ciolek) [reply Makefile]: As of now, we only have a linux/amd64 docker image. This allows us to cross-compile for darwin/amd64 and darwin/arm64. You can also pass GOOS and GOARCH to the parent Makefile from the host to cross-compile for other platforms, for example:

make GOOS=linux GOARCH=loong64 kraft

File: .goreleaser-staging.yaml (hunk 2 above)
Comment (jake-ciolek) [reply .goreleaser-staging.yaml]: Thanks

File: .goreleaser-stable.yaml (hunk 1 above)
Comment (jake-ciolek) [reply .goreleaser-stable.yaml]: Thanks. It's a dictionary so that the name of the output directory stays the same (kraft-linuxu_linux_amd64_v1)

**Comment by jake-ciolek** [comment]
@nderjung Thanks. I've modified the commit message so it only fixes #30

**Comment by jake-ciolek** [comment]
@nderjung I've added the run-related changes you mentioned.

File: Makefile (hunk 3 above)
Comment (nderjung) [reply Makefile]: Nipick: Can you please keep the space-indentation as it was before. Keeps the git blame neater and the file well-aligned. Make shouldn't be complaining about these types of indentation.

File: Makefile [hunk 6]
```diff
@@ -55,6 +58,29 @@ CMAKE       ?= cmake
 # Misc
 Q           ?= @
 
+MACHINE_OS ?= $(shell uname -s)
+MACHINE_ARCH ?= $(shell uname -m)
+GOOS ?= linux
+GOARCH ?= amd64
```
Comment (nderjung) [inline Makefile]: Can these please be indented to the same level as before and also replace all instances of `MACHINE_` with`UNAME_`.

File: machine/platform/register_darwin.go (hunk 4 above)
Comment (nderjung) [reply machine/platform/register_darwin.go:13]: Yessir!

File: Makefile (hunk 6 above)
Comment (jake-ciolek) [reply Makefile]: Done

File: Makefile (hunk 3 above)
Comment (jake-ciolek) [reply Makefile]: Done

**Review by nderjung — APPROVED** [review]
Thanks a lot!

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>
Reviewed-by: Alexander Jung <alex@unikraft.io>
Approved-by: Alexander Jung <alex@unikraft.io>

**Pull request merged**